File: python/small_context_rag_solution/retrieval.py
from __future__ import annotations

import re

from .config import (
    CHUNK_RETRIEVAL_K_PER_DOC,
    DOC_RETRIEVAL_K,
    EXPAND_NEIGHBOR_CHUNKS,
    RAW_CONTEXT_TOKEN_BUDGET,
)
from .models import ChunkRecord, HierarchicalIndex, SearchDocument
from .text import approximate_tokens, words


def retrieve_summary_path(
    question: str,
    index: HierarchicalIndex,
) -> tuple[list[SearchDocument], list[SearchDocument]]:
    document_hits = index.document_summary_store.similarity_search(
        question,
        k=min(DOC_RETRIEVAL_K, len(index.documents_by_id)),
    )

    chunk_hits = []
    seen_chunk_ids = set()

    for document_hit in document_hits:
        doc_id = str(document_hit.metadata["doc_id"])
        chunk_store = index.chunk_summary_stores_by_doc_id[doc_id]
        doc_chunk_count = len(index.chunks_by_doc_id[doc_id])
        per_doc_hits = chunk_store.similarity_search(
            question,
            k=min(CHUNK_RETRIEVAL_K_PER_DOC, doc_chunk_count),
        )

        for chunk_hit in per_doc_hits:
            chunk_id = str(chunk_hit.metadata["chunk_id"])

            if chunk_id in seen_chunk_ids:
                continue

            chunk_hits.append(chunk_hit)
            seen_chunk_ids.add(chunk_id)

    return document_hits, chunk_hits


def candidate_raw_chunks(
    chunk_hits: list[SearchDocument],
    index: HierarchicalIndex,
) -> list[ChunkRecord]:
    candidates = []
    seen_chunk_ids = set()

    for chunk_hit in chunk_hits:
        chunk = index.chunks_by_id[str(chunk_hit.metadata["chunk_id"])]
        related_chunk_ids = [chunk.chunk_id]

        if EXPAND_NEIGHBOR_CHUNKS:
            related_chunk_ids.extend([chunk.next_chunk_id, chunk.previous_chunk_id])

        for chunk_id in related_chunk_ids:
            if chunk_id is None or chunk_id in seen_chunk_ids:
                continue

            candidates.append(index.chunks_by_id[chunk_id])
            seen_chunk_ids.add(chunk_id)

    return candidates


def format_raw_chunk(chunk: ChunkRecord) -> str:
    return (
        f"Source: {chunk.source}\n"
        f"Chunk ID: {chunk.chunk_id}\n"
        f"Chunk position: {chunk.index}\n\n"
        f"{chunk.text}"
    )


def build_raw_context(
    chunk_hits: list[SearchDocument],
    index: HierarchicalIndex,
) -> tuple[str, list[tuple[ChunkRecord, int]], list[tuple[ChunkRecord, int]]]:
    included_chunks = []
    skipped_chunks = []
    used_tokens = 0

    for chunk in candidate_raw_chunks(chunk_hits, index):
        raw_context_part = format_raw_chunk(chunk)
        raw_context_tokens = approximate_tokens(raw_context_part)

        if used_tokens + raw_context_tokens > RAW_CONTEXT_TOKEN_BUDGET:
            skipped_chunks.append((chunk, raw_context_tokens))
            continue

        included_chunks.append((chunk, raw_context_tokens))
        used_tokens += raw_context_tokens

    included_chunks.sort(key=lambda item: (item[0].source, item[0].index))

    context = "\n\n---\n\n".join(
        format_raw_chunk(chunk)
        for chunk, _tokens in included_chunks
    )

    return context, included_chunks, skipped_chunks


def answer_from_raw_context(question: str, raw_context: str) -> str:
    if not raw_context:
        return "I do not know from the retrieved raw chunks."

    content_lines = []

    for line in raw_context.splitlines():
        stripped = line.strip()

        if not stripped:
            continue

        if stripped == "---":
            continue

        if stripped.startswith(("Source:", "Chunk ID:", "Chunk position:")):
            continue

        if stripped.startswith("#"):
            continue

        content_lines.append(stripped)

    content_text = " ".join(content_lines)
    question_terms = set(words(question))
    asks_about_failure = bool(question_terms & {"fail", "failure", "failures"})
    sentences = [
        sentence.strip()
        for sentence in re.split(r"(?<=[.!?])\s+", content_text)
        if sentence.strip()
    ]
    scored = []

    for sentence in sentences:
        sentence_terms = set(words(sentence))
        score = len(question_terms & sentence_terms)

        if asks_about_failure:
            lowered = sentence.lower()

            if any(
                phrase in lowered
                for phrase in (
                    "too much text",
                    "wrong chunks",
                    "too few chunks",
                    "too many chunks",
                    "miss important details",
                    "may become slower",
                )
            ):
                score += 3

        scored.append((score, sentence))

    scored.sort(key=lambda item: item[0], reverse=True)
    best = [sentence for score, sentence in scored[:3] if score > 0]

    if not best:
        return "I do not know from the retrieved raw chunks."

    return " ".join(best)


def print_retrieval_trace(
    document_hits: list[SearchDocument],
    chunk_hits: list[SearchDocument],
    included_chunks: list[tuple[ChunkRecord, int]],
    skipped_chunks: list[tuple[ChunkRecord, int]],
) -> None:
    print("\nDocument summary hits:")
    for position, hit in enumerate(document_hits, start=1):
        print(f"{position}. {hit.metadata['source']}")
        print(f"   summary: {hit.page_content}")

    print("\nChunk summary hits:")
    for position, hit in enumerate(chunk_hits, start=1):
        print(f"{position}. {hit.metadata['chunk_id']}")
        print(f"   summary: {hit.page_content}")

    print("\nRaw chunks included:")
    for position, (chunk, tokens) in enumerate(included_chunks, start=1):
        preview = " ".join(chunk.text.split())[:140]
        print(f"{position}. {chunk.chunk_id} ({tokens} approx tokens)")
        print(f"   {preview}...")

    print("\nRaw chunks skipped:")
    for position, (chunk, tokens) in enumerate(skipped_chunks, start=1):
        preview = " ".join(chunk.text.split())[:140]
        print(f"{position}. {chunk.chunk_id} ({tokens} approx tokens)")
        print(f"   {preview}...")


def answer_question(question: str, index: HierarchicalIndex) -> str:
    document_hits, chunk_hits = retrieve_summary_path(question, index)
    raw_context, included_chunks, skipped_chunks = build_raw_context(chunk_hits, index)
    print_retrieval_trace(document_hits, chunk_hits, included_chunks, skipped_chunks)
    return answer_from_raw_context(question, raw_context)

File: typescript/src/retrieval.ts
import {
  chunkRetrievalKPerDoc,
  docRetrievalK,
  expandNeighborChunks,
  rawContextTokenBudget,
} from "./config.js";
import {
  type ChunkRecord,
  type HierarchicalIndex,
  type SearchDocument,
} from "./models.js";
import { approximateTokens, words } from "./text.js";

export function retrieveSummaryPath(
  question: string,
  index: HierarchicalIndex,
): [SearchDocument[], SearchDocument[]] {
  const documentHits = index.documentSummaryStore.similaritySearch(
    question,
    Math.min(docRetrievalK, index.documentsById.size),
  );
  const chunkHits: SearchDocument[] = [];
  const seenChunkIds = new Set<string>();

  for (const documentHit of documentHits) {
    const docId = String(documentHit.metadata.docId);
    const chunkStore = index.chunkSummaryStoresByDocId.get(docId);
    const docChunks = index.chunksByDocId.get(docId);

    if (!chunkStore || !docChunks) {
      continue;
    }

    const perDocHits = chunkStore.similaritySearch(
      question,
      Math.min(chunkRetrievalKPerDoc, docChunks.length),
    );

    for (const chunkHit of perDocHits) {
      const chunkId = String(chunkHit.metadata.chunkId);

      if (seenChunkIds.has(chunkId)) {
        continue;
      }

      chunkHits.push(chunkHit);
      seenChunkIds.add(chunkId);
    }
  }

  return [documentHits, chunkHits];
}

export function candidateRawChunks(
  chunkHits: SearchDocument[],
  index: HierarchicalIndex,
): ChunkRecord[] {
  const candidates: ChunkRecord[] = [];
  const seenChunkIds = new Set<string>();

  for (const chunkHit of chunkHits) {
    const chunk = index.chunksById.get(String(chunkHit.metadata.chunkId));

    if (!chunk) {
      continue;
    }

    const relatedChunkIds: Array<string | null> = [chunk.chunkId];

    if (expandNeighborChunks) {
      relatedChunkIds.push(chunk.nextChunkId, chunk.previousChunkId);
    }

    for (const chunkId of relatedChunkIds) {
      if (chunkId === null || seenChunkIds.has(chunkId)) {
        continue;
      }

      const rawChunk = index.chunksById.get(chunkId);

      if (!rawChunk) {
        continue;
      }

      candidates.push(rawChunk);
      seenChunkIds.add(chunkId);
    }
  }

  return candidates;
}

export function formatRawChunk(chunk: ChunkRecord): string {
  return [
    `Source: ${chunk.source}`,
    `Chunk ID: ${chunk.chunkId}`,
    `Chunk position: ${chunk.index}`,
    "",
    chunk.text,
  ].join("\n");
}

export function buildRawContext(
  chunkHits: SearchDocument[],
  index: HierarchicalIndex,
): [string, Array<[ChunkRecord, number]>, Array<[ChunkRecord, number]>] {
  const includedChunks: Array<[ChunkRecord, number]> = [];
  const skippedChunks: Array<[ChunkRecord, number]> = [];
  let usedTokens = 0;

  for (const chunk of candidateRawChunks(chunkHits, index)) {
    const rawContextPart = formatRawChunk(chunk);
    const rawContextTokens = approximateTokens(rawContextPart);

    if (usedTokens + rawContextTokens > rawContextTokenBudget) {
      skippedChunks.push([chunk, rawContextTokens]);
      continue;
    }

    includedChunks.push([chunk, rawContextTokens]);
    usedTokens += rawContextTokens;
  }

  includedChunks.sort(
    ([left], [right]) => left.source.localeCompare(right.source) || left.index - right.index,
  );

  const context = includedChunks.map(([chunk]) => formatRawChunk(chunk)).join("\n\n---\n\n");
  return [context, includedChunks, skippedChunks];
}

export function answerFromRawContext(question: string, rawContext: string): string {
  if (!rawContext) {
    return "I do not know from the retrieved raw chunks.";
  }

  const contentText = rawContext
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean)
    .filter((line) => line !== "---")
    .filter((line) => !line.startsWith("Source:"))
    .filter((line) => !line.startsWith("Chunk ID:"))
    .filter((line) => !line.startsWith("Chunk position:"))
    .filter((line) => !line.startsWith("#"))
    .join(" ");

  const questionTerms = new Set(words(question));
  const asksAboutFailure = ["fail", "failure", "failures"].some((word) =>
    questionTerms.has(word),
  );
  const sentences = contentText
    .split(/(?<=[.!?])\s+/)
    .map((sentence) => sentence.trim())
    .filter(Boolean);
  const scored = sentences
    .map((sentence) => {
      let score = words(sentence).filter((word) => questionTerms.has(word)).length;

      if (asksAboutFailure) {
        const lowered = sentence.toLowerCase();

        if (
          [
            "too much text",
            "wrong chunks",
            "too few chunks",
            "too many chunks",
            "miss important details",
            "may become slower",
          ].some((phrase) => lowered.includes(phrase))
        ) {
          score += 3;
        }
      }

      return {
        score,
        sentence,
      };
    })
    .filter((item) => item.score > 0)
    .sort((left, right) => right.score - left.score)
    .slice(0, 3);

  if (scored.length === 0) {
    return "I do not know from the retrieved raw chunks.";
  }

  return scored.map((item) => item.sentence).join(" ");
}

export function printRetrievalTrace(
  documentHits: SearchDocument[],
  chunkHits: SearchDocument[],
  includedChunks: Array<[ChunkRecord, number]>,
  skippedChunks: Array<[ChunkRecord, number]>,
): void {
  console.log("\nDocument summary hits:");
  documentHits.forEach((hit, index) => {
    console.log(`${index + 1}. ${hit.metadata.source}`);
    console.log(`   summary: ${hit.pageContent}`);
  });

  console.log("\nChunk summary hits:");
  chunkHits.forEach((hit, index) => {
    console.log(`${index + 1}. ${hit.metadata.chunkId}`);
    console.log(`   summary: ${hit.pageContent}`);
  });

  console.log("\nRaw chunks included:");
  includedChunks.forEach(([chunk, tokens], index) => {
    const preview = chunk.text.replace(/\s+/g, " ").slice(0, 140);
    console.log(`${index + 1}. ${chunk.chunkId} (${tokens} approx tokens)`);
    console.log(`   ${preview}...`);
  });

  console.log("\nRaw chunks skipped:");
  skippedChunks.forEach(([chunk, tokens], index) => {
    const preview = chunk.text.replace(/\s+/g, " ").slice(0, 140);
    console.log(`${index + 1}. ${chunk.chunkId} (${tokens} approx tokens)`);
    console.log(`   ${preview}...`);
  });
}

export function answerQuestion(question: string, index: HierarchicalIndex): string {
  const [documentHits, chunkHits] = retrieveSummaryPath(question, index);
  const [rawContext, includedChunks, skippedChunks] = buildRawContext(chunkHits, index);
  printRetrievalTrace(documentHits, chunkHits, includedChunks, skippedChunks);
  return answerFromRawContext(question, rawContext);
}
